Dynamic fee adjustment during high network activity (#585)
ava-labs/builders-hub: docs/build/tutorials/platform/adjusting-gas-price-during-high-network-activity.md

---

> description: This tutorial will help users to adjust their priority fee and max fee cap during high network activity and take the benefits of dynamic fees algorithm.

Reviewer: ```suggestion
description: This tutorial will help users to adjust their priority fee and max fee cap during high network activity and take advantage of the benefits of dynamic fees.
```

---

> # Adjusting gas price during high network activity

Reviewer 2: Please change all titles to Camel case.

## Good to Know Keywords and Concepts

The amount of computation used by a transaction is measured in units of `gas`. Each gas has a dynamic price (to be paid in AVAX) which depends upon the network activity, and this is known as `gas price`.

To avoid draining the user's wallet due to using computation infinitely, transactions need to be submitted with a `gas limit`, which denotes the maximum units of gas that a particular transaction could consume.

If a transaction uses more than the gas limit, then it would be rejected. Total fees paid by the user is calculated as `(gas consumed) * (gas price)`, and is known as `gas fees`. Similarly, maximum gas fees are calculated as `(gas limit) * (gas price)`.

Earlier users have to bid how much they want to pay per gas for their transactions. The transaction with a higher bid was picked by the validators. This made the estimation difficult, and users often pay a higher price than required.

This was later replaced by a more deterministic algorithm to estimate the gas price according to the network activity.

> Each block has a target for the units of gas to be included in it. This is called `TargetGas`. Currently, it is `15M` gas units per block. To make the gas price more predictable, each block is associated with a `base price or base fee` for each gas.

Author: Thanks! I was confused on this part and will update the section.

---
description: This tutorial will help users to adjust their priority fee and max fee cap during high network activity and take advantage of the benefits of dynamic fee transactions.
---

# Adjusting Gas Price During High Network Activity

> Sometimes during periods of high network activity, transactions either remain pending for a very long duration or instantly get a failed transaction notification. This may cause panic among users if they can't figure out what went wrong!

Reviewer 3: Can we change 'This may cause panic among users' to something a bit less alarming, like 'This may confuse and frustrate users, especially if they don't understand why their transactions are not getting accepted.'

Author: Sounds good!

## Probable Reasons You are Here

* Your transaction has stalled, and you don't know what to do
* Your transaction has failed, with an error - `transaction underpriced`
* It's your first transaction, and you want to be sure about any potential issues
* Just for general knowledge on adjusting dynamic fee settings

If these are your reasons for being here, then you can either go through this entire section, for a better understanding of the scenario or directly skip to the [solution](#adjusting-gas-fees-before-submitting-the-transaction).

## Good to Know Keywords and Concepts

The amount of computation used by a transaction is measured in units of `gas`. Each unit of gas is paid for in AVAX at the `gas price` for the transaction. The `gas price` of the transaction is determined by the parameters of the transaction and the `base fee` of the block that it is included in.

> To avoid draining the user's wallet due to using computation infinitely, transactions are submitted with a `gas limit`, which denotes the maximum units of gas that a particular transaction is allowed to consume.

Reviewer: ```suggestion
To avoid draining the user's wallet due to non-terminating execution through the EVM, transactions are submitted with a `gas limit`, which denotes the maximum units of gas that a particular transaction is allowed to consume.
```

```

Consider the following examples (here GWEI or nAVAX is one-billionth of AVAX) -

| Transaction | Max Fee Cap | Gas tip cap | Base price | Effective tip | Total price |
| ----------- | ----------- | ----------- | ---------- | ------------- | ----------- |
| 1           | 50 GWEI     | 0 GWEI      | 25 GWEI    | 0 GWEI        | 25 GWEI     |
| 2           | 50 GWEI     | 0 GWEI      | 50 GWEI    | 0 GWEI        | 50 GWEI     |
| 3           | 50 GWEI     | 0 GWEI      | 60 GWEI    | 0 GWEI        | **PENDING**|
| A           | 50 GWEI     | 10 GWEI     | 40 GWEI    | 10 GWEI       | 50 GWEI     |
| B           | 40 GWEI     | 40 GWEI     | 40 GWEI    | 0 GWEI        | 40 GWEI     |

Look at transactions **A** and **B** (the bottom two transactions). In these scenarios, it looks like transaction B is paying a higher tip, however, this depends on the base fee of the block where the transactions are included. The effective tip of A is more than that of B. So, if both of these transaction competes for being included in the next block, then the validators would prioritize transaction A since it pays a higher effective tip.

## Why my Transaction is on Hold or Failing?

> If your transaction is failing and giving an error - `transaction underpriced`, then the max fee cap of your transaction must be less than the minimum base price that the network supports (as of now, it's 25 nAVAX or GWEI). Although the base fee is automatically estimated in wallets like Metamask, you can try increasing the max fee cap in the wallet.

Reviewer: ```suggestion
If your transaction is failing and giving an error - `transaction underpriced`, then the max fee cap of your transaction must be less than the minimum base price that the network supports (as of now, it's 25 nAVAX or GWEI). Although the base fee is automatically estimated in wallets like MetaMask, you can try increasing the max fee cap in the wallet.
```

> During a period of heavy congestion on the network, all submitted transactions can't be included in the same block, due to the block's gas limit. So, validators choose transactions giving higher priority to transactions with the highest effective tips. Your transaction will have to wait until the effective tip is highest among the pending transactions.

Reviewer: ```suggestion
During a period of heavy congestion on the network, all submitted transactions can't be included in the same block, due to the block's gas limit. So, validators choose transactions giving higher priority to transactions with the highest effective tips.
```

> Another reason for pending transactions is the max fee cap being below the current base fee that the network is charging. In this case, you nat need to increase the max fee cap of the transaction.

Reviewer: ```suggestion
Another reason your transaction may get stuck in pending, is that the max fee cap may be below the current base fee that the network is charging. In this case, you need to increase the max fee cap of your transaction above the current base fee for it to be included in the block.
```

## Adjusting Gas Fees Before Submitting the Transaction

You may not need to edit the gas fees on normal days. This is only required if there is heavy congestion on the network, and the base fees are frequently fluctuating.

1. Let's create a sample transaction on Avalanche Mainnet, in which we will be sending 0.1 AVAX to a receiver using the MetaMask. By clicking **Next** we can review gas fees and the amount which we want to send.

![](/img/dynamic-fees-adjustment-1.png)

2. On the review page, you can see the estimated priority and max fee for this transaction. Now click on **EDIT**, to adjust these fees according to network requirements.

![](/img/dynamic-fees-adjustment-2.png)

3. On this page, you can edit the priority fee (gas tip cap) and max fee (max fee cap) using the advanced options or you can use the MetaMask's inbuilt gas estimation. You can estimate the max fee as shown on [snowtrace](https://snowtrace.io/gastracker) which represents the average max fee over the last 3 seconds. For more detailed statistics, you can have a look [here](https://stats.avax.network/dashboard/c-chain-activity/).

![](/img/dynamic-fees-adjustment-3.png)

> 4. If the network activity is high, you have to edit the priority and max fees accordingly, as given on snowtrace. Consider the example below, where the average max fee is 30 GWEI (nAVAX).

Reviewer: Let's get rid of this reference to snowtrace's gastracker. This is not necessary and we should recommend that users simply set the max fee cap to what they are willing to pay and set the tip cap what they are willing to pay above the base fee.

Let's recommend 100gwei and 10gwei as examples/defaults and make it clear that if their transaction is stuck, they can check snowtrace to see the base fee of the last block and increase the fees on their transaction if they are willing to pay at the higher rate.
